File: app/src/main/java/com/dpis/module/PerAppDisplayConfigSource.java
package com.dpis.module;

import com.dpis.module.fonts.FontApplyMode;

import com.dpis.module.fonts.hookdomain.FontHookDomainDecision;

import com.dpis.module.runtime.systemserver.PerAppDisplayConfig;

import com.dpis.module.viewport.ViewportPropertyBridge;
import com.dpis.module.viewport.ViewportApplyMode;
import com.dpis.module.viewport.ViewportTargetSpec;

import android.os.SystemClock;

import java.util.Set;
import java.util.function.Supplier;

public final class PerAppDisplayConfigSource {
    public interface SnapshotProvider {
        ConfigSnapshot get();
    }

    public interface PackageFallbackProvider {
        PackageConfigSnapshot get(String packageName);
    }

    public static final class RefreshingSnapshotProvider implements SnapshotProvider {
        private static final long FAILURE_LOG_MIN_INTERVAL_MILLIS = 30_000L;

        interface Clock {
            long nowMillis();
        }

        private final Supplier<ConfigSnapshot> loader;
        private final long ttlMillis;
        private final Clock clock;
        private volatile ConfigSnapshot snapshot;
        private volatile long loadedAtMillis;
        private volatile long lastRefreshAttemptAtMillis = Long.MIN_VALUE;
        private volatile long lastFailureLoggedAtMillis;
        private volatile String lastFailureMessage;

        public RefreshingSnapshotProvider(
                Supplier<ConfigSnapshot> loader,
                long ttlMillis) {
            this(loader, ttlMillis, SystemClock::elapsedRealtime);
        }

        RefreshingSnapshotProvider(
                Supplier<ConfigSnapshot> loader,
                long ttlMillis,
                Clock clock) {
            this.loader = loader;
            this.ttlMillis = Math.max(0L, ttlMillis);
            this.clock = clock != null ? clock : System::currentTimeMillis;
        }

        @Override
        public ConfigSnapshot get() {
            long now = clock.nowMillis();
            ConfigSnapshot current = snapshot;
            if (!shouldRefresh(current, now)) {
                return current;
            }
            synchronized (this) {
                now = clock.nowMillis();
                current = snapshot;
                if (!shouldRefresh(current, now)) {
                    return current;
                }
                try {
                    lastRefreshAttemptAtMillis = now;
                    ConfigSnapshot loaded = loader != null ? loader.get() : null;
                    snapshot = loaded != null ? loaded : ConfigSnapshot.empty();
                    loadedAtMillis = now;
                    lastFailureMessage = null;
                    return snapshot;
                } catch (Throwable throwable) {
                    logFailureIfNeeded(throwable, now);
                    return current != null ? current : ConfigSnapshot.empty();
                }
            }
        }

        private boolean shouldRefresh(ConfigSnapshot current, long nowMillis) {
            if (ttlMillis == 0L) {
                return true;
            }
            if (current != null) {
                return (nowMillis - loadedAtMillis) >= ttlMillis;
            }
            return lastRefreshAttemptAtMillis == Long.MIN_VALUE
                    || (nowMillis - lastRefreshAttemptAtMillis) >= ttlMillis;
        }

        private void logFailureIfNeeded(Throwable throwable, long nowMillis) {
            String message = throwable.getClass().getName()
                    + ": "
                    + throwable.getMessage();
            if (message.equals(lastFailureMessage)
                    && (nowMillis - lastFailureLoggedAtMillis)
                    < FAILURE_LOG_MIN_INTERVAL_MILLIS) {
                return;
            }
            lastFailureMessage = message;
            lastFailureLoggedAtMillis = nowMillis;
            DpisLog.e("config snapshot refresh failed: " + message, throwable);
        }
    }

    private final SnapshotProvider snapshotProvider;
    private final PackageFallbackProvider packageFallbackProvider;

    public PerAppDisplayConfigSource(DpisConfigStore store) {
        this(ConfigSnapshotLoader.fromStore(store));
    }

    public PerAppDisplayConfigSource(ConfigSnapshot snapshot) {
        this(() -> snapshot);
    }

    public PerAppDisplayConfigSource(SnapshotProvider snapshotProvider) {
        this(snapshotProvider, null);
    }

    public PerAppDisplayConfigSource(SnapshotProvider snapshotProvider,
                              PackageFallbackProvider packageFallbackProvider) {
        this.snapshotProvider = snapshotProvider;
        this.packageFallbackProvider = packageFallbackProvider;
    }

    public static PerAppDisplayConfigSource withLegacyRuntimePropertyFallback(
            SnapshotProvider snapshotProvider) {
        return new PerAppDisplayConfigSource(
                snapshotProvider,
                PerAppDisplayConfigSource::loadLegacyRuntimePropertyConfig);
    }

    public PerAppDisplayConfig get(String packageName) {
        ConfigSnapshot snapshot = getSnapshot();
        PackageConfigSnapshot packageConfig = snapshot.getPackage(packageName);
        if (packageConfig != null && !packageConfig.dpisEnabled) {
            return null;
        }
        PackageConfigSnapshot fallbackConfig = getFallbackPackageConfig(packageName);
        if (fallbackConfig != null) {
            packageConfig = fallbackConfig;
        }
        if (packageConfig == null || !packageConfig.dpisEnabled) {
            return null;
        }
        // Runtime viewport properties are intentionally projected per package so
        // app-process changes can apply without rebuilding all configured packages.
        ViewportTargetSpec runtimeTargetSpec = ViewportPropertyBridge.readTargetSpec(packageName);
        ViewportTargetSpec targetViewportSpec = runtimeTargetSpec.isEnabled()
                ? runtimeTargetSpec
                : packageConfig.targetViewportSpec;
        String viewportMode = ViewportApplyMode.normalize(packageConfig.targetViewportMode);
        if (ViewportApplyMode.COMPAT.equals(viewportMode) || ViewportApplyMode.OFF.equals(viewportMode)) {
            targetViewportSpec = ViewportTargetSpec.off();
        }
        Integer targetFontScalePercent = packageConfig.targetFontScalePercent;
        String targetFontMode = packageConfig.targetFontMode;
        boolean fontConfigured = FontApplyMode.isEnabled(targetFontMode)
                && targetFontScalePercent != null;
        if (!targetViewportSpec.isEnabled() && !fontConfigured) {
            return null;
        }
        boolean hyperOsNativeFlutterEnabled = FontHookDomainDecision
                .isHyperOsNativeFlutterEnabled(snapshot, packageConfig);
        return new PerAppDisplayConfig(packageName, targetViewportSpec, viewportMode,
                targetFontScalePercent, targetFontMode,
                hyperOsNativeFlutterEnabled,
                packageConfig.hookDomainOverride);
    }

    public Set<String> getConfiguredPackages() {
        return getSnapshot().getConfiguredPackages();
    }

    public boolean isSystemServerHooksEnabled() {
        return getSnapshot().isSystemServerHooksEnabled();
    }

    private ConfigSnapshot getSnapshot() {
        ConfigSnapshot snapshot = snapshotProvider != null ? snapshotProvider.get() : null;
        return snapshot != null ? snapshot : ConfigSnapshot.empty();
    }

    private PackageConfigSnapshot getFallbackPackageConfig(String packageName) {
        if (packageFallbackProvider == null || packageName == null || packageName.isBlank()) {
            return null;
        }
        return packageFallbackProvider.get(packageName);
    }

    private static PackageConfigSnapshot loadLegacyRuntimePropertyConfig(String packageName) {
        ConfigSnapshot snapshot = ConfigSnapshotLoader.fromStore(
                new DpisConfigStore(new RuntimePropertyConfigPreferences(packageName)));
        return snapshot.getPackage(packageName);
    }
}

File: app/src/main/java/com/dpis/module/config/PerAppDisplayConfigSource.kt
package com.dpis.module.config

import android.os.SystemClock
import com.dpis.module.DpisConfigStore
import com.dpis.module.DpisLog
import com.dpis.module.fonts.FontApplyMode
import com.dpis.module.fonts.hookdomain.FontHookDomainDecision
import com.dpis.module.runtime.systemserver.PerAppDisplayConfig
import com.dpis.module.viewport.ViewportApplyMode
import com.dpis.module.viewport.ViewportPropertyBridge
import com.dpis.module.viewport.ViewportTargetSpec
import java.util.function.Supplier

class PerAppDisplayConfigSource(
    private val snapshotProvider: SnapshotProvider?,
    private val packageFallbackProvider: PackageFallbackProvider?,
) {
    fun interface SnapshotProvider {
        fun get(): ConfigSnapshot?
    }

    fun interface PackageFallbackProvider {
        fun get(packageName: String): PackageConfigSnapshot?
    }

    constructor(store: DpisConfigStore?) : this(ConfigSnapshotLoader.fromStore(store))
    constructor(snapshot: ConfigSnapshot?) : this(SnapshotProvider { snapshot })
    constructor(snapshotProvider: SnapshotProvider?) : this(snapshotProvider, null)

    fun get(packageName: String?): PerAppDisplayConfig? {
        val snapshot = getSnapshot()
        var packageConfig = snapshot.getPackage(packageName)
        if (packageConfig != null && !packageConfig.dpisEnabled) return null
        getFallbackPackageConfig(packageName)?.let { packageConfig = it }
        if (packageConfig == null || !packageConfig.dpisEnabled) return null

        val runtimeTargetSpec = ViewportPropertyBridge.readTargetSpec(packageName)
        var targetViewportSpec =
            if (runtimeTargetSpec.isEnabled()) runtimeTargetSpec else packageConfig.targetViewportSpec
        val viewportMode = ViewportApplyMode.normalize(packageConfig.targetViewportMode)
        if (viewportMode == ViewportApplyMode.COMPAT || viewportMode == ViewportApplyMode.OFF) {
            targetViewportSpec = ViewportTargetSpec.off()
        }
        val fontConfigured = FontApplyMode.isEnabled(packageConfig.targetFontMode) &&
            packageConfig.targetFontScalePercent != null
        if (!targetViewportSpec.isEnabled() && !fontConfigured) return null

        val hyperOsNativeFlutterEnabled = FontHookDomainDecision
            .isHyperOsNativeFlutterEnabled(snapshot, packageConfig)
        return PerAppDisplayConfig(
            packageName,
            targetViewportSpec,
            viewportMode,
            packageConfig.targetFontScalePercent,
            packageConfig.targetFontMode,
            hyperOsNativeFlutterEnabled,
            packageConfig.hookDomainOverride,
        )
    }

    fun getConfiguredPackages(): MutableSet<String?> = getSnapshot().getConfiguredPackages().toMutableSet()
    fun isSystemServerHooksEnabled(): Boolean = getSnapshot().isSystemServerHooksEnabled()

    private fun getSnapshot(): ConfigSnapshot =
        snapshotProvider?.get() ?: ConfigSnapshot.empty()

    private fun getFallbackPackageConfig(packageName: String?): PackageConfigSnapshot? =
        if (packageFallbackProvider == null || packageName.isNullOrBlank()) {
            null
        } else {
            packageFallbackProvider.get(packageName)
        }

    companion object {
        @JvmStatic
        fun withLegacyRuntimePropertyFallback(
            snapshotProvider: SnapshotProvider?,
        ) = PerAppDisplayConfigSource(
            snapshotProvider,
            PackageFallbackProvider { packageName ->
                ConfigSnapshotLoader.fromStore(
                    DpisConfigStore(RuntimePropertyConfigPreferences(packageName)),
                ).getPackage(packageName)
            },
        )
    }

    class RefreshingSnapshotProvider @JvmOverloads constructor(
        private val loader: Supplier<ConfigSnapshot>?,
        ttlMillis: Long,
        private val clock: Clock = Clock { SystemClock.elapsedRealtime() },
    ) : SnapshotProvider {
        fun interface Clock {
            fun nowMillis(): Long
        }

        private val ttlMillis = ttlMillis.coerceAtLeast(0L)
        @Volatile private var snapshot: ConfigSnapshot? = null
        @Volatile private var loadedAtMillis = 0L
        @Volatile private var lastRefreshAttemptAtMillis = Long.MIN_VALUE
        @Volatile private var lastFailureLoggedAtMillis = 0L
        @Volatile private var lastFailureMessage: String? = null

        override fun get(): ConfigSnapshot? {
            var now = clock.nowMillis()
            var current = snapshot
            if (!shouldRefresh(current, now)) return current
            synchronized(this) {
                now = clock.nowMillis()
                current = snapshot
                if (!shouldRefresh(current, now)) return current
                return try {
                    lastRefreshAttemptAtMillis = now
                    snapshot = loader?.get() ?: ConfigSnapshot.empty()
                    loadedAtMillis = now
                    lastFailureMessage = null
                    snapshot
                } catch (throwable: Throwable) {
                    logFailureIfNeeded(throwable, now)
                    current ?: ConfigSnapshot.empty()
                }
            }
        }

        private fun shouldRefresh(current: ConfigSnapshot?, now: Long): Boolean =
            if (ttlMillis == 0L) true
            else if (current != null) now - loadedAtMillis >= ttlMillis
            else lastRefreshAttemptAtMillis == Long.MIN_VALUE ||
                now - lastRefreshAttemptAtMillis >= ttlMillis

        private fun logFailureIfNeeded(throwable: Throwable, now: Long) {
            val message = "${throwable.javaClass.name}: ${throwable.message}"
            if (message == lastFailureMessage &&
                now - lastFailureLoggedAtMillis < FAILURE_LOG_MIN_INTERVAL_MILLIS
            ) return
            lastFailureMessage = message
            lastFailureLoggedAtMillis = now
            DpisLog.e("config snapshot refresh failed: $message", throwable)
        }

        companion object {
            private const val FAILURE_LOG_MIN_INTERVAL_MILLIS = 30_000L
        }
    }
}
